cli/main.c

```diff
@@ -1313,6 +1316,66 @@ static int fw_info(int argc, char **argv)
 	return 0;
 }
 
+static int compare_device_secure_version(struct switchtec_dev *dev, int img_fd,
+				         int assume_yes_ver)
+{
+	int ret;
+	int prompt = 0;
+	uint32_t dev_sec_ver;
+	struct switchtec_fw_image_info info;
+	struct switchtec_sn_ver_info sn_info = {};
+
+	if (assume_yes_ver || switchtec_is_gen3(dev))
+		return 0;
+
+	ret = switchtec_fw_file_info(img_fd, &info);
+	if (ret)
+		return ret;
+
+	ret = switchtec_sn_ver_get(dev, &sn_info);
+	if (ret) {
+		sn_info.ver_bl2 = 0xffffffff;
+		sn_info.ver_main = 0xffffffff;
+		sn_info.ver_km = 0xffffffff;
+	}
+
+	switch (info.type) {
+	case SWITCHTEC_FW_TYPE_BL2:
+		if (info.secure_version > sn_info.ver_bl2) {
+			prompt = 1;
+			dev_sec_ver = sn_info.ver_bl2;
+		}
+		break;
+	case SWITCHTEC_FW_TYPE_IMG:
+		if (info.secure_version > sn_info.ver_main) {
+			prompt = 1;
+			dev_sec_ver = sn_info.ver_main;
+		}
+		break;
+	case SWITCHTEC_FW_TYPE_KEY:
+		if (info.secure_version > sn_info.ver_km) {
+			prompt = 1;
+			dev_sec_ver = sn_info.ver_km;
+		}
+		break;
+	default:
+		break;
+	}
+
+	if ((prompt == 0) || assume_yes_ver)
+		return 0;
+
+	fprintf(stderr, "\n\nWARNING:\n"
```

Comment:
I'd put the check outside the static function. This function, based on it's name, just does the compare... That also means you don't have to pass it an extra variable.

Reply:
I'd still put this outside the function in the CLI's function. It's significantly more reusable to have a function that returns a property than a static function prints a specific message for its specific call site.

Reply:
Also, the warning should really be printed regardless of whether assume_yes is set.... assume_yes is really just for not asking to continue, -- it's not "don't print warnings"...